File: dotnet/tests/PIIMasker.Tests/ModelRollbackTests.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.IO;
using System.Threading;
using System.Threading.Tasks;
using Xunit;

namespace PIIMasker.Tests;

/// <summary>
/// Probation, the launch counter, and the revert. The failure this guards against is silent — a
/// model that loads fine and quietly stops detecting PII — so every branch is driven with an
/// injected probe rather than inferred from a real one.
/// </summary>
public sealed class ModelRollbackTests : IDisposable
{
    private readonly string scratch = Path.Combine(
        Path.GetTempPath(), "piimasker-rollback-" + Guid.NewGuid().ToString("N"));

    public ModelRollbackTests() => Directory.CreateDirectory(scratch);

    public void Dispose()
    {
        try { Directory.Delete(scratch, recursive: true); }
        catch (IOException) { }
        catch (UnauthorizedAccessException) { }
    }

    private string InstallRoot => scratch;

    /// <summary>An in-memory store, so a test never touches the machine's real settings.</summary>
    private sealed class MemoryStore : IRollbackStore
    {
        private readonly Dictionary<string, string> values = new(StringComparer.Ordinal);
        public string? Read(string key) => values.TryGetValue(key, out var v) ? v : null;
        public void Write(string key, string? value)
        {
            if (value == null) values.Remove(key); else values[key] = value;
        }
    }

    private ModelRollback NewRollback(IRollbackStore? store = null) =>
        new(InstallRoot, store ?? new MemoryStore());

    private void InstallVersion(string version)
    {
        var dir = Path.Combine(InstallRoot, version);
        Directory.CreateDirectory(dir);
        foreach (var name in ModelInstaller.RequiredModelFiles)
            File.WriteAllText(Path.Combine(dir, name), version);
        ModelInstaller.PointCurrent(version, InstallRoot);
    }

    /// <summary>
    /// Derives <c>Found</c> from the real anchor list rather than taking it separately, so a probe
    /// result here cannot describe a state the production probe could never produce — which is how
    /// a "healthy" fixture ends up failing the two-of-three pass mark.
    /// </summary>
    private static Func<string, CancellationToken, Task<ModelProbeResult>> Probe(
        bool ran, params string[] missed) =>
        (_, _) => Task.FromResult(new ModelProbeResult(
            ran,
            PrivacyFilter.ProbeAnchors.Where(a => !missed.Contains(a, StringComparer.Ordinal)).ToList(),
            missed));

    private static readonly Func<string, CancellationToken, Task<ModelProbeResult>> Healthy =
        Probe(ran: true);

    private static readonly Func<string, CancellationToken, Task<ModelProbeResult>> DetectsNothing =
        Probe(ran: true, [.. PrivacyFilter.ProbeAnchors]);

    [Fact]
    public void ArmingNeedsSomethingToGoBackTo()
    {
        InstallVersion("2026.08.1");
        var rollback = NewRollback();

        // No previous version on disk and not a known pin: a failure here is a reinstall, not a
        // rollback, and arming would promise a revert that cannot happen.
        rollback.Arm("2026.09.1", "never-published");
        Assert.False(rollback.IsPending);

        // A same-version repair has no previous version at all.
        rollback.Arm("2026.08.1", "2026.08.1");
        Assert.False(rollback.IsPending);

        // The previous version's bytes are on disk, so the revert is a pointer flip.
        rollback.Arm("2026.09.1", "2026.08.1");
        Assert.True(rollback.IsPending);
        Assert.Equal(("2026.09.1", "2026.08.1"), rollback.Probation);
    }

    /// <summary>
    /// Two launches, not one: an unrelated crash during launch would otherwise blocklist a model
    /// that was never at fault.
    /// </summary>
    [Fact]
    public void AProbationSurvivesOneUnrelatedCrashAndIsRevertedAfterTheSecond()
    {
        InstallVersion("2026.08.1");
        InstallVersion("2026.09.1");
        var rollback = NewRollback();
        rollback.Arm("2026.09.1", "2026.08.1");

        Assert.Equal(ModelRollback.LaunchOutcome.Proving, rollback.ResolveAtLaunch());
        Assert.Equal(ModelRollback.LaunchOutcome.Proving, rollback.ResolveAtLaunch());
        Assert.Equal(ModelRollback.LaunchOutcome.RolledBack, rollback.ResolveAtLaunch());

        Assert.Equal("2026.08.1", new ModelInstaller(InstallRoot).InstalledVersion());
        Assert.Contains("2026.09.1", rollback.BlockedVersions);
        Assert.False(rollback.IsPending);
    }

    [Fact]
    public async Task AHealthyModelIsConfirmedAndNothingIsBlocked()
    {
        InstallVersion("2026.08.1");
        InstallVersion("2026.09.1");
        var rollback = NewRollback();
        rollback.Arm("2026.09.1", "2026.08.1");

        var verdict = await rollback.VerifyAsync(Healthy);

        Assert.Equal(ModelRollback.Verdict.Verified, verdict);
        Assert.False(rollback.IsPending);
        Assert.Empty(rollback.BlockedVersions);
        Assert.Equal("2026.09.1", new ModelInstaller(InstallRoot).InstalledVersion());
    }

    /// <summary>
    /// The failure the whole mechanism exists for: the model loads perfectly and detects nothing, so
    /// the fail-closed gate stops closing and unmasked text starts flowing.
    /// </summary>
    [Fact]
    public async Task AModelThatLoadsButDetectsNothingIsRevertedAndRefused()
    {
        InstallVersion("2026.08.1");
        InstallVersion("2026.09.1");
        var rollback = NewRollback();
        rollback.Arm("2026.09.1", "2026.08.1");

        var verdict = await rollback.VerifyAsync(DetectsNothing);

        Assert.Equal(ModelRollback.Verdict.Reverted, verdict);
        Assert.Equal("2026.08.1", new ModelInstaller(InstallRoot).InstalledVersion());
        Assert.Contains("2026.09.1", rollback.BlockedVersions);
        Assert.False(rollback.IsPending);
    }

    /// <summary>
    /// A probe that could not run against a complete directory IS a verdict: the files were all
    /// there, they were opened, and they were not a model.
    /// </summary>
    [Fact]
    public async Task AProbeThatCannotRunAgainstACompleteStoreIsAVerdict()
    {
        InstallVersion("2026.08.1");
        InstallVersion("2026.09.1");
        var rollback = NewRollback();
        rollback.Arm("2026.09.1", "2026.08.1");

        var verdict = await rollback.VerifyAsync(Probe(ran: false));

        Assert.Equal(ModelRollback.Verdict.Reverted, verdict);
        Assert.Contains("2026.09.1", rollback.BlockedVersions);
    }

    /// <summary>
    /// A probe that could not run against an INCOMPLETE store says nothing about the model.
    /// Blocklisting there would pin the machine to an older version over a one-off hiccup.
    /// </summary>
    [Fact]
    public async Task AProbeThatCannotRunAgainstAnIncompleteStoreIsInconclusive()
    {
        InstallVersion("2026.08.1");
        InstallVersion("2026.09.1");
        var rollback = NewRollback();
        rollback.Arm("2026.09.1", "2026.08.1");
        File.Delete(Path.Combine(InstallRoot, "2026.09.1", "tokenizer.json"));

        var verdict = await rollback.VerifyAsync(Probe(ran: false));

        Assert.Equal(ModelRollback.Verdict.Inconclusive, verdict);
        Assert.Empty(rollback.BlockedVersions);
        // Left armed, so the next launch tries again and the counter eventually resolves it.
        Assert.True(rollback.IsPending);
    }

    /// <summary>
    /// An install landing mid-probe moves the target. Applying the verdict then would blocklist a
    /// version the probe never looked at.
    /// </summary>
    [Fact]
    public async Task AVerdictIsDiscardedWhenAnInstallMovesTheTargetMidProbe()
    {
        InstallVersion("2026.08.1");
        InstallVersion("2026.09.1");
        var rollback = NewRollback();
        rollback.Arm("2026.09.1", "2026.08.1");

        var verdict = await rollback.VerifyAsync((_, _) =>
        {
            // A newer install re-arms probation while the probe is out.
            rollback.Arm("2026.10.1", "2026.09.1");
            return Task.FromResult(new ModelProbeResult(true, [], ["email address"]));
        });

        Assert.Equal(ModelRollback.Verdict.Idle, verdict);
        Assert.Empty(rollback.BlockedVersions);
        Assert.Equal(("2026.10.1", "2026.09.1"), rollback.Probation);
    }

    /// <summary>
    /// Reverting to bytes that are gone would replace a working-but-suspect model with nothing at
    /// all — strictly worse than what is being recovered from. When the missing version is still a
    /// known pin it can be fetched, so probation is LEFT ARMED for
    /// <see cref="ModelRollback.RevertInstallingIfNeededAsync"/> rather than abandoned.
    /// </summary>
    [Fact]
    public void ARevertWaitsForADownloadWhenTheMissingVersionIsStillAKnownPin()
    {
        var known = ModelPin.Current.Version;
        InstallVersion(known);
        InstallVersion("2026.09.1");
        var rollback = NewRollback();
        rollback.Arm("2026.09.1", known);
        Directory.Delete(Path.Combine(InstallRoot, known), recursive: true);

        var reverted = rollback.Revert();

        Assert.False(reverted);
        Assert.Equal("2026.09.1", new ModelInstaller(InstallRoot).InstalledVersion());
        Assert.NotNull(rollback.PendingRevertDownload);
        Assert.True(rollback.IsPending);
        // Nothing is blocked yet: the version on probation has not been judged, only deferred.
        Assert.Empty(rollback.BlockedVersions);
    }

    /// <summary>
    /// When the previous version's bytes are gone AND no pin can bring it back, probation is cleared
    /// rather than retried forever — there is no recovery to wait for.
    /// </summary>
    [Fact]
    public void ARevertIsAbandonedWhenNothingCanBringThePreviousVersionBack()
    {
        InstallVersion("2020.01.1");
        InstallVersion("2026.09.1");
        var rollback = NewRollback();
        rollback.Arm("2026.09.1", "2020.01.1");
        Directory.Delete(Path.Combine(InstallRoot, "2020.01.1"), recursive: true);

        var reverted = rollback.Revert();

        Assert.False(reverted);
        Assert.Null(rollback.PendingRevertDownload);
        Assert.False(rollback.IsPending);
        Assert.Equal("2026.09.1", new ModelInstaller(InstallRoot).InstalledVersion());
    }

    [Fact]
    public void ClearingTheBlocklistLetsARefusedVersionBeTriedAgain()
    {
        var rollback = NewRollback();
        rollback.Block("2026.09.1");
        Assert.Contains("2026.09.1", rollback.BlockedVersions);

        rollback.ClearBlocklist();

        Assert.Empty(rollback.BlockedVersions);
    }

    [Fact]
    public void BlockingIsIdempotent()
    {
        var rollback = NewRollback();
        rollback.Block("2026.09.1");
        rollback.Block("2026.09.1");

        Assert.Single(rollback.BlockedVersions);
    }

    [Fact]
    public async Task NothingArmedMeansNothingToVerify()
    {
        Assert.Equal(ModelRollback.Verdict.Idle, await NewRollback().VerifyAsync(DetectsNothing));
        Assert.Equal(ModelRollback.LaunchOutcome.None, NewRollback().ResolveAtLaunch());
    }

    /// <summary>
    /// The file-backed store has to survive a process boundary, because the launch counter is
    /// meaningless if the count resets on every start.
    /// </summary>
    [Fact]
    public void TheFileStoreSurvivesANewInstanceAndRoundTripsTheAttemptCount()
    {
        InstallVersion("2026.08.1");
        InstallVersion("2026.09.1");
        var path = Path.Combine(scratch, "rollback.json");

        var first = new ModelRollback(InstallRoot, new FileRollbackStore(path));
        first.Arm("2026.09.1", "2026.08.1");
        Assert.Equal(ModelRollback.LaunchOutcome.Proving, first.ResolveAtLaunch());

        // A second instance reading the same file is the next launch.
        var second = new ModelRollback(InstallRoot, new FileRollbackStore(path));
        Assert.Equal(("2026.09.1", "2026.08.1"), second.Probation);
        Assert.Equal(ModelRollback.LaunchOutcome.Proving, second.ResolveAtLaunch());
        Assert.Equal(ModelRollback.LaunchOutcome.RolledBack, second.ResolveAtLaunch());
    }

    /// <summary>
    /// A corrupt record must not stop the host starting. The cost is one unproved model losing its
    /// probation, which the launch counter catches; throwing on the launch path would be worse.
    /// </summary>
    [Fact]
    public void ACorruptRecordReadsAsNoRecordRatherThanThrowing()
    {
        var path = Path.Combine(scratch, "rollback.json");
        File.WriteAllText(path, "{ this is not json");
        var rollback = new ModelRollback(InstallRoot, new FileRollbackStore(path));

        Assert.False(rollback.IsPending);
        Assert.Empty(rollback.BlockedVersions);
        Assert.Equal(ModelRollback.LaunchOutcome.None, rollback.ResolveAtLaunch());
    }
}

File: swift/Tests/PIIMaskerTests/ModelRollbackTests.swift
import Foundation
import Testing

@testable import PIIMasker

/// Hermetic tests for `ModelRollback`: fake versioned model dirs under a temp install root and a
/// scratch defaults suite. No model, no network.
///
/// The probe half — does the *real* model still find PII — is in `ModelProbeTests`, which needs
/// weights on disk.
@Suite("Model rollback")
struct ModelRollbackTests {

    /// A version dir holding every file the loader opens — enough for `isCompleteModelDir`, which is
    /// what the revert checks before flipping back to it.
    @discardableResult
    static func installVersion(_ version: String, marker: String, into root: URL) throws -> URL {
        let dir = root.appendingPathComponent(version, isDirectory: true)
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        try marker.data(using: .utf8)!.write(to: dir.appendingPathComponent("model.onnx"))
        for f in ModelInstaller.requiredModelFiles where f != "model.onnx" {
            try "{}".data(using: .utf8)!.write(to: dir.appendingPathComponent(f))
        }
        return dir
    }

    static func currentVersion(in root: URL) -> String? {
        try? FileManager.default.destinationOfSymbolicLink(
            atPath: root.appendingPathComponent("current").path)
    }

    /// A fresh install root plus its own defaults suite, so tests never see each other's records.
    static func fixture() throws -> (install: URL, defaults: UserDefaults, rollback: ModelRollback, trash: Trash) {
        let root = TestPublish.scratch()
        let install = root.appendingPathComponent("install", isDirectory: true)
        try FileManager.default.createDirectory(at: install, withIntermediateDirectories: true)
        let suite = "com.github.pii-masker.tests-\(UUID().uuidString)"
        let defaults = UserDefaults(suiteName: suite)!
        return (install, defaults, ModelRollback(installRoot: install, defaults: defaults), Trash(root))
    }

    // MARK: - Arming

    /// Arming needs somewhere to go back to: a first install and a same-version repair do not
    /// qualify, so a failure there is a reinstall rather than a rollback.
    @Test func armingNeedsSomewhereToGoBackTo() throws {
        let f = try Self.fixture()
        try Self.installVersion("v1", marker: "model-v1", into: f.install)
        try ModelInstaller.pointCurrent(at: "v1", in: f.install)

        f.rollback.arm(new: "v1", previous: "v1")
        #expect(!f.rollback.isPending)

        f.rollback.arm(new: "v2", previous: "gone")
        #expect(!f.rollback.isPending)
        _ = f.trash
    }

    /// A real version change arms, and only the launch counter advances until a verdict lands.
    @Test func aVersionChangeArmsAndCountsLaunches() throws {
        let f = try Self.fixture()
        try Self.installVersion("v1", marker: "model-v1", into: f.install)
        try Self.installVersion("v2", marker: "model-v2", into: f.install)
        try ModelInstaller.pointCurrent(at: "v2", in: f.install)

        f.rollback.arm(new: "v2", previous: "v1")
        #expect(f.rollback.isPending)
        #expect(f.rollback.probation?.new == "v2")
        #expect(f.rollback.probation?.previous == "v1")
        #expect(f.rollback.resolveAtLaunch() == .proving)
        #expect(f.rollback.resolveAtLaunch() == .proving)
        _ = f.trash
    }

    /// Out of launches: revert without probing. This is the crash case — nothing in-process ever
    /// gets to report a verdict, so the launch counter is the only signal there is.
    @Test func outOfLaunchesRevertsUnprobed() throws {
        let f = try Self.fixture()
        try Self.installVersion("v1", marker: "model-v1", into: f.install)
        try Self.installVersion("v2", marker: "model-v2", into: f.install)
        try ModelInstaller.pointCurrent(at: "v2", in: f.install)
        f.rollback.arm(new: "v2", previous: "v1")
        _ = f.rollback.resolveAtLaunch()
        _ = f.rollback.resolveAtLaunch()

        #expect(f.rollback.resolveAtLaunch() == .rolledBack(from: "v2"))
        #expect(Self.currentVersion(in: f.install) == "v1")
        #expect(f.rollback.blockedVersions == ["v2"])
        #expect(!f.rollback.isPending)
        #expect(f.rollback.resolveAtLaunch() == .none)
        _ = f.trash
    }

    // MARK: - The blocklist

    /// A blocked version is not reinstalled over a working model…
    @Test func aBlockedVersionIsSkippedOverAWorkingModel() async throws {
        let f = try Self.fixture()
        let dist = f.install.deletingLastPathComponent().appendingPathComponent("dist")
        try FileManager.default.createDirectory(at: dist, withIntermediateDirectories: true)
        try Self.installVersion("v1", marker: "model-v1", into: f.install)
        try ModelInstaller.pointCurrent(at: "v1", in: f.install)
        f.rollback.block("v2")

        let pin = try TestPublish.publish(version: "v2", marker: "model-v2", into: dist)
        let skipped = try await ModelInstaller(pin: pin, installRoot: f.install)
            .ensureLatest(skipping: f.rollback.blockedVersions)
        #expect(!skipped.installed)
        #expect(Self.currentVersion(in: f.install) == "v1")

        // …and an explicit retry clears the refusal: the user outranks our verdict, and otherwise
        // the only pinned version stays refused forever with no way to say "try again".
        f.rollback.clearBlocklist()
        #expect(f.rollback.blockedVersions.isEmpty)
        let retried = try await ModelInstaller(pin: pin, installRoot: f.install)
            .ensureLatest(skipping: f.rollback.blockedVersions)
        #expect(retried.installed)
        #expect(Self.currentVersion(in: f.install) == "v2")
        _ = f.trash
    }

    // MARK: - Reverting

    /// A revert with nowhere to go back to must not strand the host on no model at all — and the
    /// record is dropped rather than retried forever.
    @Test func aRevertWithNowhereToGoIsAbandoned() throws {
        let f = try Self.fixture()
        try Self.installVersion("v2", marker: "model-v2", into: f.install)
        try ModelInstaller.pointCurrent(at: "v2", in: f.install)
        // Plant the record directly: `arm` would refuse, which is the previous assertion.
        f.defaults.set(
            ["new": "v2", "previous": "v1"],
            forKey: ModelRollback.SettingsKeys.default.probation)

        #expect(!f.rollback.revert())
        #expect(Self.currentVersion(in: f.install) == "v2")
        #expect(!f.rollback.isPending)
        _ = f.trash
    }

    /// The second revert mode. When the previous version's bytes are gone but the pin list still
    /// names it, the record is LEFT ARMED so it can be fetched rather than abandoned — the branch
    /// that becomes reachable the first time a second model version ships.
    @Test func aRevertToAKnownPinIsHeldForDownload() throws {
        let f = try Self.fixture()
        let known = ModelPin.current.version
        try Self.installVersion("next", marker: "model-next", into: f.install)
        try ModelInstaller.pointCurrent(at: "next", in: f.install)
        f.defaults.set(
            ["new": "next", "previous": known],
            forKey: ModelRollback.SettingsKeys.default.probation)

        #expect(f.rollback.pendingRevertDownload?.version == known)
        #expect(!f.rollback.revert())
        // Still armed: unlike the vanished-version case, this one has a recovery.
        #expect(f.rollback.isPending)
        #expect(Self.currentVersion(in: f.install) == "next")
        _ = f.trash
    }

    /// …and `arm` accepts such a pair, where it would refuse a version nothing can bring back.
    @Test func armAcceptsAKnownPinNotOnDisk() throws {
        let f = try Self.fixture()
        f.rollback.arm(new: "next", previous: ModelPin.current.version)
        #expect(f.rollback.isPending)

        f.rollback.confirm()
        f.rollback.arm(new: "next", previous: "never-published")
        #expect(!f.rollback.isPending)
        _ = f.trash
    }

    // MARK: - Verdicts

    /// A probe that could not run for reasons that say nothing about the model — the store not
    /// resolving, an install repointing `current` underneath it — must not blocklist a version. Only
    /// the loader having opened the files and found they were not a model is a verdict.
    @Test func anUnprobeableModelIsInconclusive() async throws {
        let f = try Self.fixture()
        try Self.installVersion("m1", marker: "x", into: f.install)
        try ModelInstaller.pointCurrent(at: "m1", in: f.install)
        f.rollback.arm(new: "m2", previous: "m1")
        try FileManager.default.removeItem(at: f.install.appendingPathComponent("current"))

        let verdict = await f.rollback.verify { _ in
            PrivacyFilter.ProbeResult(found: [], missed: PrivacyFilter.probeAnchors, ran: false)
        }
        #expect(verdict == .inconclusive, "\(verdict)")
        #expect(f.rollback.isPending)
        #expect(f.rollback.blockedVersions.isEmpty)
        _ = f.trash
    }

    /// THE failure this exists for: a model that loads perfectly and quietly detects less. No crash,
    /// no error — the fail-closed gate simply stops closing and PII flows onward. It cannot be built
    /// on demand (it needs a valid graph that under-detects), so the probe is handed in.
    @Test func aModelThatUnderDetectsIsReverted() async throws {
        let f = try Self.fixture()
        try Self.installVersion("m8", marker: "masks-properly", into: f.install)
        try Self.installVersion("m9", marker: "under-detecting", into: f.install)
        try ModelInstaller.pointCurrent(at: "m9", in: f.install)
        f.rollback.arm(new: "m9", previous: "m8")
        #expect(f.rollback.isPending)

        let verdict = await f.rollback.verify { _ in
            PrivacyFilter.ProbeResult(
                found: [PrivacyFilter.probeAnchors[0]],
                missed: Array(PrivacyFilter.probeAnchors.dropFirst()), ran: true)
        }
        #expect(verdict == .reverted, "\(verdict)")
        #expect(Self.currentVersion(in: f.install) == "m8")
        #expect(f.rollback.blockedVersions.contains("m9"))
        _ = f.trash
    }

    /// Two of three is model jitter, not a broken model — that one is kept.
    @Test func twoOfThreeAnchorsStillPasses() async throws {
        let f = try Self.fixture()
        try Self.installVersion("m8", marker: "masks-properly", into: f.install)
        try Self.installVersion("m9", marker: "fine", into: f.install)
        try ModelInstaller.pointCurrent(at: "m9", in: f.install)
        f.rollback.arm(new: "m9", previous: "m8")

        let verdict = await f.rollback.verify { _ in
            PrivacyFilter.ProbeResult(
                found: Array(PrivacyFilter.probeAnchors.prefix(2)),
                missed: [PrivacyFilter.probeAnchors[2]], ran: true)
        }
        #expect(verdict == .verified, "\(verdict)")
        #expect(Self.currentVersion(in: f.install) == "m9")
        #expect(!f.rollback.blockedVersions.contains("m9"))
        #expect(!f.rollback.isPending)
        _ = f.trash
    }

    /// Nothing armed is not a verdict about anything.
    @Test func verifyWithNothingArmedIsIdle() async throws {
        let f = try Self.fixture()
        let verdict = await f.rollback.verify { _ in
            PrivacyFilter.ProbeResult(found: [], missed: PrivacyFilter.probeAnchors, ran: true)
        }
        #expect(verdict == .idle, "\(verdict)")
        _ = f.trash
    }

    /// The keys are the caller's, so a host can keep its own namespace — and an existing host can
    /// keep reading records it wrote under its old key.
    @Test func settingsKeysAreInjectable() throws {
        let f = try Self.fixture()
        let keys = ModelRollback.SettingsKeys(probation: "custom.probation", failedVersions: "custom.failed")
        let custom = ModelRollback(installRoot: f.install, defaults: f.defaults, keys: keys)
        try Self.installVersion("a", marker: "a", into: f.install)
        custom.arm(new: "b", previous: "a")
        custom.block("b")

        #expect(f.defaults.dictionary(forKey: "custom.probation") != nil)
        #expect(f.defaults.stringArray(forKey: "custom.failed") == ["b"])
        // The default-keyed view sees none of it.
        #expect(!f.rollback.isPending)
        #expect(f.rollback.blockedVersions.isEmpty)
        _ = f.trash
    }
}
